### PR TITLE
Adsense Datapoint Refactoring

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #453 

## Relevant technical choices

* Restores ability for a data request to pass a specific `dateRange`. If not provided, it will fallback to the filter hook provided value.
* Updated `AdSensePerformanceWidget` to use `withData` rather than executing batch requests internally (allows use of `toState`)

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by aaemnnosttv** [comment]
Everything passing except VRT https://travis-ci.com/google/site-kit-wp/builds/126940714

File: assets/js/components/data.js [hunk 1]
```diff
@@ -59,6 +59,28 @@ const lazilySetupLocalCache = () => {
...
+	 *
+	 * @param String The selected date range. Default 'Last 28 days'.
+	 */
+	const dateRangeSlug = stringToSlug( applyFilters( 'googlesitekit.dateRange', __( 'Last 28 days', 'google-site-kit' ) ) );
+	// Make copies for reference safety, ensuring data exists.
+	const request = { data: {}, ...originalRequest };
+	// Use the dateRange in request.data if passed, fallback to filter-provided value.
+	request.data = { dateRange: dateRangeSlug, ...request.data };
+
+	return request;
```
Comment (felixarntz) [inline assets/js/components/data.js]: Nice!

File: assets/js/components/data.js [hunk 2]
```diff
@@ -59,6 +59,28 @@ const lazilySetupLocalCache = () => {
...
+ * @param {Object} originalRequest Data request object.
+ * @return {Object} New data request object.
+ */
+const requestWithDateRange = ( originalRequest ) => {
+	/**
+	 * Filter the date range used for queries.
+	 *
+	 * @param String The selected date range. Default 'Last 28 days'.
+	 */
+	const dateRangeSlug = stringToSlug( applyFilters( 'googlesitekit.dateRange', __( 'Last 28 days', 'google-site-kit' ) ) );
```
Comment (felixarntz) [inline assets/js/components/data.js]: This filter is now triggered for every `request` object. Can you change it so that it still triggers once like before, and the value is passed to this helper function instead?

File: includes/Modules/AdSense.php [hunk 3]
```diff
@@ -522,102 +525,43 @@ function( \Google_Service_AdSense_Alert $alert ) {
...
-							break;
-						case 'last-90-days':
-							$daysago = 90;
-							break;
-						case 'last-28-days':
-						default:
-							$daysago = 28;
-							break;
+
+					if ( empty( $data['dateRange'] ) ) {
```
Comment (felixarntz) [inline includes/Modules/AdSense.php]: Can we instead have a default of `last-28-days` here, like most (all?) of the other datapoints do?

File: includes/Modules/AdSense.php [hunk 4]
```diff
@@ -731,25 +685,90 @@ protected function parse_data_response( $method, $datapoint, $response ) {
...
+				return array(
+					date( 'Y-m-d', strtotime( 'yesterday' ) ),
+					date( 'Y-m-d', strtotime( 'yesterday' ) ),
+				);
+			case 'same-day-last-week':
+				return array(
+					date( 'Y-m-d', strtotime( '7 days ago' ) ),
+					date( 'Y-m-d', strtotime( '7 days ago' ) ),
+				);
+			case '7-days':
```
Comment (felixarntz) [inline includes/Modules/AdSense.php]: Shouldn't we make this the same as `last-7-days`? Having this extra clause is kind of pointless and confusing. If you call `$this->parse_date_range( $date_range )`, you get exactly this.

File: includes/Modules/AdSense.php [hunk 5]
```diff
@@ -731,25 +685,90 @@ protected function parse_data_response( $method, $datapoint, $response ) {
...
+				return array(
+					date( 'Y-m-d', strtotime( '7 days ago' ) ),
+					date( 'Y-m-d', strtotime( 'yesterday' ) ),
+				);
+			case 'prev-7-days':
+				return array(
+					date( 'Y-m-d', strtotime( '14 days ago' ) ),
+					date( 'Y-m-d', strtotime( '8 days ago' ) ),
+				);
```
Comment (felixarntz) [inline includes/Modules/AdSense.php]: This theoretically can also be accomplished with `$this->parse_date_range( $date_range, 1, 8 )`, but let's keep this as is because it's a special case. We need to revisit the whole date range parsing anyway.

File: includes/Modules/AdSense.php [hunk 6]
```diff
@@ -731,25 +685,90 @@ protected function parse_data_response( $method, $datapoint, $response ) {
...
+			// Intentional fallthrough.
+			case 'last-7-days':
+			case 'last-14-days':
+			case 'last-28-days':
+			case 'last-90-days':
+				preg_match( '/last-(?P<days>\d+)-days/', $date_range, $matches );
+
+				return array(
+					date( 'Y-m-d', strtotime( "{$matches['days']} days ago" ) ),
+					date( 'Y-m-d', strtotime( 'today' ) ),
```
Comment (felixarntz) [inline includes/Modules/AdSense.php]: I'd prefer if we didn't redo this here and instead had a `default` clause which would simply call `$this->parse_date_range( $date_range )` (see above).

File: includes/Modules/AdSense.php [hunk 7]
```diff
@@ -731,25 +685,90 @@ protected function parse_data_response( $method, $datapoint, $response ) {
...
+				);
+			case 'this-month-last-year':
+				$last_year          = intval( date( 'Y' ) ) - 1;
+				$last_date_of_month = date( 't', strtotime( $last_year . '-' . date( 'm' ) . '-01' ) );
+
+				return array(
+					date( $last_year . '-m-01' ),
+					date( $last_year . '-m-' . $last_date_of_month ),
+				);
+			case '28-days':
```
Comment (felixarntz) [inline includes/Modules/AdSense.php]: Same as above, this should instead become (and re-use) `last-28-days`.

File: includes/Modules/AdSense.php (hunk 3 above)
Comment (aaemnnosttv) [reply includes/Modules/AdSense.php]: Yes, I forgot we decided that this was not a required parameter 👍

File: includes/Modules/AdSense.php (hunk 4 above)
Comment (aaemnnosttv) [reply includes/Modules/AdSense.php]: I noticed that, but the two are actually slightly different and the `earnings` datapoint doesn't use `parse_date_range` (nor did it before the refactor - perhaps for this reason?)

`last-7-days` end date is `today` while `7-days` uses `yesterday`. I'm not sure why this is but that's how it was, so I kept it. If this is safe to change then just let me know!

File: includes/Modules/AdSense.php (hunk 5 above)
Comment (aaemnnosttv) [reply includes/Modules/AdSense.php]: Agreed. FWIW I think what's here is more readable than `parse_date_range( $date_range, 1, 8 )`. It would be nice to come up with something more concise without sacrificing readability.

File: includes/Modules/AdSense.php (hunk 7 above)
Comment (aaemnnosttv) [reply includes/Modules/AdSense.php]: As above, this one uses the alternate end date. Ok to change?

File: assets/js/components/data.js (hunk 2 above)
Comment (aaemnnosttv) [reply assets/js/components/data.js]: Updated 👍

File: includes/Modules/AdSense.php (hunk 6 above)
Comment (aaemnnosttv) [reply includes/Modules/AdSense.php]: Updated, although not using `default` to preserve error behavior if invalid/unknown range passed. We could make an exception for any `last-*-days` though?

File: includes/Modules/AdSense.php (hunk 7 above)
Comment (felixarntz) [reply includes/Modules/AdSense.php]: They should definitely become one (for both 7 and 28 days). There is no point in having two different versions. Whether we make them 7/28 days ago until yesterday or 6/27 days ago until today depends on whether the AdSense API even exposes data as recent as the last few hours. The only reason we don't include today in other modules is that there wouldn't be data anyway. @adamsilverstein Do you have any information about that?

In other words, it should either be `parse_date_range( $date_range, 1, 1 )` or `parse_date_range( $date_range, 1, 0 )`.

**Comment by adamsilverstein** [comment]
Seeing some data issues when testing this PR:

image
